radix-engine/src/ledger/traits.rs

```diff
@@ -33,110 +33,78 @@ pub trait QueryableSubstateStore {
 
 /// A ledger stores all transactions and substates.
 pub trait SubstateStore {
-    /// Top Level Objects
-    fn get_resource_def(&self, resource_def_id: ResourceDefId) -> Option<ResourceDef>;
-    fn put_resource_def(&mut self, resource_def_id: ResourceDefId, resource_def: ResourceDef);
-    fn get_package(&self, package_id: PackageId) -> Option<Package>;
-    fn put_package(&mut self, package_id: PackageId, package: Package);
-    fn get_component(&self, component_id: ComponentId) -> Option<Component>;
-    fn put_component(&mut self, component_id: ComponentId, component: Component);
-
-    /// Child Objects
-    fn get_lazy_map_entry(
-        &self,
-        component_id: ComponentId,
-        lazy_map_id: &LazyMapId,
-        key: &[u8],
-    ) -> Option<Vec<u8>>;
-    fn put_lazy_map_entry(
-        &mut self,
-        component_id: ComponentId,
-        lazy_map_id: LazyMapId,
-        key: Vec<u8>,
-        value: Vec<u8>,
-    );
-    fn get_vault(&self, component_id: ComponentId, vault_id: &VaultId) -> Vault;
-    fn put_vault(&mut self, component_id: ComponentId, vault_id: VaultId, vault: Vault);
-    fn get_non_fungible(
-        &self,
-        resource_def_id: ResourceDefId,
-        key: &NonFungibleKey,
-    ) -> Option<NonFungible>;
-    fn put_non_fungible(
-        &mut self,
-        resource_def_id: ResourceDefId,
-        key: &NonFungibleKey,
-        non_fungible: NonFungible,
-    );
+    fn get_substate<T: Encode>(&self, address: &T) -> Option<Vec<u8>>;
```

Comment:
Maybe we can introduce a trait for the three "address" types?

Reply:
Would prefer not to overdesign this part now since it'll be subject to a lot of refactoring soon.

Reply:
Sounds good!